ouroboros-consensus/src/Ouroboros/Consensus/Protocol/WithEBBs.hs

```diff
@@ -48,13 +63,77 @@ instance (OuroborosTag p) => OuroborosTag (WithEBBs p) where
 
   preferCandidate       (WithEBBNodeConfig cfg) = preferCandidate       cfg
   compareCandidates     (WithEBBNodeConfig cfg) = compareCandidates     cfg
-  checkIsLeader         (WithEBBNodeConfig cfg) = checkIsLeader         cfg
-  rewindChainState      (WithEBBNodeConfig cfg) = rewindChainState      cfg
   protocolSecurityParam (WithEBBNodeConfig cfg) = protocolSecurityParam cfg
 
   applyChainState (WithEBBNodeConfig cfg) lv b cs = case
     eitherHeaderOrEbb cfg b of
-      Right hdr -> applyChainState @p cfg lv hdr cs
+      Right hdr ->
+          wrap <$> applyChainState @p cfg lv hdr (underlyingChainState cs)
+        where
+          wrap x = ChainStateWithEBBs
+            { firstNonEBBSlot      = Just $ case firstNonEBBSlot cs of
+                Just s  -> s
+                Nothing -> blockSlotOfHeader cfg [b] hdr
+            , underlyingChainState = x
+            }
       -- Currently the hash chain is maintained in the ledger state, so we just
       -- ignore the EBB for protocol concerns.
       Left _ -> return cs
+
+  checkIsLeader (WithEBBNodeConfig cfg) slot lv cs =
+    checkIsLeader cfg slot lv (underlyingChainState cs)
+
+  rewindChainState (WithEBBNodeConfig cfg) cs mSlot =
+      wrap <$> rewindChainState cfg (underlyingChainState cs) mSlot'
+    where
+      effectivelyRewindingToOrigin = case mSlot of
+        At slot
+          | Just s <- firstNonEBBSlot cs -> slot < s
+        _                                -> True
+
+      (mSlot', firstNonEBBSlot')
+         | effectivelyRewindingToOrigin = (Origin, Nothing)
+         | otherwise                    = (mSlot, firstNonEBBSlot cs)
+
+      wrap x = ChainStateWithEBBs
+        { firstNonEBBSlot      = firstNonEBBSlot'
+        , underlyingChainState = x
+        }
+
+-- | A wrapper around @'ChainState' p@ with a little extra bookeeping data.
+--
+data ChainStateWithEBBs p = ChainStateWithEBBs
+  { firstNonEBBSlot      :: !(Maybe SlotNo)
+    -- ^ earliest slot incorporated into 'underlyingChainState' that is not an
+    -- EBB
+  , underlyingChainState :: !(ChainState p)
+  }
+
+deriving instance Show (ChainState p) => Show (ChainStateWithEBBs p)
+
+-- | Create a 'ChainStateWithEBBs' from a 'ChainState' that does not include
+-- any non-EBB blocks.
+--
+initChainStateWithEBBs :: ChainState p -> ChainStateWithEBBs p
+initChainStateWithEBBs x = ChainStateWithEBBs
+  { firstNonEBBSlot = Nothing
+  , underlyingChainState = x
+  }
+
+encodeChainStateWithEBBs ::
+    (ChainState p -> CBOR.Encoding)
+  -> ChainStateWithEBBs p
+  -> CBOR.Encoding
+encodeChainStateWithEBBs f (ChainStateWithEBBs x0 x1) =
+  CBOR.encodeListLen 2
+  <> CBOR.encode x0
+  <> f x1
+
+decodeChainStateWithEBBs ::
+     CBOR.Decoder s (ChainState p)
+  -> CBOR.Decoder s (ChainStateWithEBBs p)
+decodeChainStateWithEBBs f = do
+  CBOR.decodeListLenOf 2
+  !x0 <- CBOR.decode
+  !x1 <- f
```

Comment:
For a general monad, it could make a difference (forcing `x0` before `f` instead of after it). Off the top of my head, for the CPS style monad here, I'm not sure if it could make a difference -- I'm wondering if it depends on the top-level continuation?

Regardless of if it could make a difference, I would try benchmarks to determine if it does. Assuming that's more effort than this PR currently justifies: should I be pessimistic (include bangs) or optimistic (no bangs at all)? Is there a established team precedent here?

The other occurrences of `decodeListLenOf` in the repo don't use any bangs at all, so I'll go with that for now.

Reply:
Let's stick to no bangs for now. We still need to do more profiling and checking whether the state is in normal formal anyway.